### PR TITLE
Added 2 new endpoints for PlaylistsApi.

### PR DESCRIPTION
**Pull request by akshays2112**

### CODE REVIEW

**Comment by akshays2112** [comment]
Hi Daniel,

Requesting just a quick look at the way I solved the Dynamic Object/Class Type problem with Newtonsoft.Json ShouldSerialize<PropName>() method.

This is WIP.  Please approve of this and then I continue and finish of all endpoints first pass.  Then consolidate using your classes and return types (classes) from the other pull request of yours and that should save you time I think.  Well other than review time.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by DanielLarsenNZ** [comment]
Hi Akshay,

You don't need `ShouldSerialize` at all. You can replace this object with my example here: https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR7

Notice the use of `NullValueHandling` attribute arguments here.

I have also changed `SpotifyWebApi` to return a new `SpotifyResponse` instead of `HttpResponseMessage` from `Post` methods. I have added a new generic overload so that you can return a deserialized object from `Post<T>()`. I have updated your PlaylistApi implementation to use this new helper.

**Comment by DanielLarsenNZ** [comment]
For the release notes:

This release changes the return types of `Post()` and `Put()` from `HttpResponseMessage` to `SpotifyResponse<T>`. No internal callers were using the return type so this is not a breaking change. A new generic `Post<T>()` internal helper has been added to deserialize POST response bodies to an object typed `T`

`Description` is added to `PlaylistSimplified`

File: src/SpotifyApi.NetCore/Models/Playlist.cs [hunk 1]
```diff
@@ -269,4 +269,16 @@ public partial class PlaylistPaged
         [JsonProperty("total")]
         public int Total { get; set; }
     }
+
+    /// <summary>
+    /// Upon adding items to a playlist a snapshot_id is returned.
+    /// </summary>
+    public partial class PlaylistSnapshotID
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/Models/Playlist.cs:282]: Please create a separate class file for this and rename to `ModifyPlaylistResponse`. See this example: https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-3070e73922144bda6f13360673a6810cR5

File: src/SpotifyApi.NetCore/IPlaylistsApi.cs [hunk 2]
```diff
@@ -118,5 +118,67 @@ public interface IPlaylistsApi
...
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PlaylistSnapshotID"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/playlists/add-tracks-to-playlist/
+        /// </remarks>
+        Task<T> AddItemsToPlaylist<T>(
+            string playlistId,
+            string[] uris,
+            int? position = null,
+            string accessToken = null
+            );
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IPlaylistsApi.cs]: We don't need a generic method for this one, thanks.

File: src/SpotifyApi.NetCore/IPlaylistsApi.cs [hunk 3]
```diff
@@ -118,5 +118,67 @@ public interface IPlaylistsApi
...
+        /// <remarks>
+        /// At least one optional parameter must be supplied.
+        /// https://developer.spotify.com/documentation/web-api/reference/playlists/change-playlist-details/
+        /// </remarks>
+        Task ChangePlaylistDetails(
+            string playlistId,
+            string name = null,
+            bool? makePublic = null,
+            bool? collaborative = null,
+            string description = null,
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IPlaylistsApi.cs]: Please use a request DTO here, see https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-bccd16b25f81b67901713ae0c33510e0R152

File: src/SpotifyApi.NetCore/PlaylistsApi.cs [hunk 4]
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
...
+        /// <param name="position">Optional. The position to insert the items, a zero-based index. For example, to insert the items in the first position: position=0 ; to insert the items in the third position: position=2. If omitted, the items will be appended to the playlist. Items are added in the order they appear in the uris array. For example: {"uris": ["spotify:track:4iV5W9uYEdYUVa79Axb7Rh","spotify:track:1301WleyT98MSxVHPZCA6M", "spotify:episode:512ojhOuo1ktJprKbVcKyQ"], "position": 3}</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PlaylistSnapshotID"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/playlists/add-tracks-to-playlist/
+        /// </remarks>
+        public async Task<T> AddItemsToPlaylist<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PlaylistsApi.cs]: Generic method is not needed here. Should return `ModifyPlaylistResponse`

File: src/SpotifyApi.NetCore/PlaylistsApi.cs [hunk 5]
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
...
+                     ArgumentException("The position if supplied has to be 0 or greater than 0.");
+
+            var builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/tracks");
+            return (await Post<T>(builder.Uri, new { uris, position }, accessToken)).Data;
+        }
+
+        #endregion
+
+        #region ChangePlaylistDetails
+        class PutDataPayloadPlaylistDetails
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PlaylistsApi.cs]: Please don't nest classed. Create a new class. See https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR5

File: src/SpotifyApi.NetCore/PlaylistsApi.cs [hunk 6]
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
...
+
+            public bool ShouldSerializeCollaborative ()
+            {
+                return Collaborative != null;
+            }
+
+            public bool ShouldSerializeDescription()
+            {
+                return !string.IsNullOrEmpty(Description);
+            }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PlaylistsApi.cs]: Don't do this. See https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR7

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]

**Comment by akshays2112** [comment]
Hi Daniel,

Super. ♥ all of it. Will do it. Drinking right now (11:30PM here in Inja) so will do it tomorrow when sober.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Review by DanielLarsenNZ — COMMENTED** [review]
How's this Akshay?

**Comment by akshays2112** [comment]
Hi Daniel,

You are an awesome coder my friend.  Yes, basically that damn deserializing bug was due to it being Paged<T> and not a regular thing.  Rest of all the changes are awesome.  Yes, this time I read through each and every one.

I think we can merge it if you are ok with it.
Then you can merge in your other pull request as you deem fit as usual.
Then I can sync up my stuff and continue closing out the remaining few endpoints easily.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by akshays2112** [comment]
Hi Daniel,

 

Went through all your changes. 100% approved.  Please merge and close out all pull requests and then I will sync and continue soon as I get word from you.  Ya the life issues were well life problems not health problems LOL.  Meh!! Sh*t happens you know LOL!

 

Cheers,

 

Akshay Srinivasan

+91-9884967112

Mumbai, India

**Pull request merged**